src/interfaces/chain.h

```diff
@@ -213,6 +213,42 @@ class Chain
     //! Calculate mempool ancestor and descendant counts for the given transaction.
     virtual void getTransactionAncestry(const uint256& txid, size_t& ancestors, size_t& descendants, size_t* ancestorsize = nullptr, CAmount* ancestorfees = nullptr) = 0;
 
+    //! For each outpoint, calculate the fee-bumping cost to spend this outpoint at the specified
+    //  feerate, including bumping its ancestors. For example, if the target feerate is 10sat/vbyte
+    //  and this outpoint refers to a mempool transaction at 5sat/vbyte, the bump fee includes the
+    //  cost to bump the mempool transaction to 10sat/vbyte (i.e. 5 * mempooltx.vsize). If that
+    //  transaction also has, say, an unconfirmed parent with a feerate of 1sat/vbyte, the bump fee
+    //  includes the cost to bump the parent (i.e. 9 * parentmempooltx.vsize).
+    //
+    //  If the outpoint comes from an unconfirmed transaction that is already above the target
+    //  feerate or bumped by its descendant(s) already, it does not need to be bumped. Its bump fee
+    //  is 0. Likewise, if any of the transaction's ancestors are already bumped, they are not
+    //  included in the transaction's bump fee.
+    //
+    //  This includes fee-bumping in RBFs. If an outpoint conflicts with another transaction in the
+    //  mempool, it is assumed that the goal is to replace that transaction. As such, the
+    //  calculation will exclude the to-be-replaced transaction, but will include the fee-bumping
+    //  cost. If bump fees of descendants of the to-be-replaced transaction are requested, the value
+    //  will be 0. Fee-related RBF rules are not included as they are logically distinct.
```

Comment:
Would it be safer not to make this assumption? The cluster would be larger, but since the algorithms are efficient, the performance difference shouldn't be a problem. The `MiniMiner` constructor would be simpler too. Even if currently we're never given a outpoint that conflicts with a transaction _not_ being replaced, could that possibly change in the future?

Initially, I thought that was I'm suggesting here wouldn't work, because suppose that the transaction that our outpoint refers to (let's call it the parent) has a low feerate, but the transaction we're replacing (the parent's existing child) has a very high feerate. If, contrary to what the code currently does, we ignore the fact that we're replacing the child (we keep the child in the cluster, as if we're spending a different output of the parent), then we might conclude that we don't need to bump the parent's fee -- _whereas we actually do_, because the child will no longer exist!

But I think this won't happen because rule 6 of our [replacement policy](https://github.com/bitcoin/bitcoin/blob/master/doc/policy/mempool-replacements.md) requires the replacement transaction to have a higher feerate than all that it's replacing.

> 6. The replacement transaction's feerate is greater than the feerates of all directly conflicting transactions.

So it seems like even if we (mistakenly) think that the existing child will remain and that it will bump the parent somewhat, it won't bump it enough to meet our requested feerate, so we'll still bump it (as required by our desired feerate).

Reply:
I’m not sure I follow your train of thought here. I think you have since reviewed this further, do you think this still needs to be addressed?

Reply:
Thinking more about this, I think the assumption is valid and I don't expect it to change soon. It is also documented here, so if it changes, people should notice the conflict.